### ISSUE TITLE
remove deprecated `std::result_of`

### ISSUE THREAD

**Issue by BenWibking** [parthenon-hpc-lab/parthenon#1008]
`src/tasks/tasks.hpp` uses `std::result_of`, which is deprecated in C++17 and removed in C++20. This causes lots of warnings on my machine:

```
In file included from /Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:30:
/Users/benwibking/athenapk/external/parthenon/src/tasks/thread_pool.hpp:121:36: warning: 'result_of<(lambda at /Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:420:20) ()>' is deprecated [-Wdeprecated-declarations]
    using return_t = typename std::result_of<F(Args...)>::type;
                                   ^
/Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:420:12: note: in instantiation of function template specialization 'parthenon::ThreadPool::enqueue<(lambda at /Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:420:20)>' requested here
      pool.enqueue([t, &ProcessTask]() { return ProcessTask(t); });
           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__type_traits/result_of.h:24:34: note: 'result_of<(lambda at /Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:420:20) ()>' has been explicitly marked deprecated here
template <class _Callable> class _LIBCPP_DEPRECATED_IN_CXX17 result_of;
                                 ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__config:862:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__config:835:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
4 warnings generated.
```

**Comment by Yurlungur** [comment]
Ping @jdolence

**Comment by pbrady** [comment]
Unless I'm misunderstanding something, this is a straightforward change to `std::invoke_result_t`.  Is this more about dropping c++14?

**Comment by BenWibking** [comment]
Simply substituting `std::invoke_result` does not work:
```
[2/150] Building CXX object src/CMakeFiles/parthenon.dir/bvals/comms/build_boundary_buffers.cpp.o
FAILED: src/CMakeFiles/parthenon.dir/bvals/comms/build_boundary_buffers.cpp.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DFS_HEADER="<filesystem>" -DFS_NAMESPACE=std::filesystem -DH5_BUILT_AS_DYNAMIC_LIB -DKOKKOS_DEPENDENCE -DVTKMDIY_MPI_AS_LIB -DVTKMDIY_NO_THREADS -I/Users/benwibking/parthenon/src -I/Users/benwibking/parthenon/build/src/generated -I/opt/homebrew/Cellar/hdf5-mpi/1.14.3/include -I/opt/homebrew/opt/libaec/include -I/Users/benwibking/parthenon/build/Kokkos -I/Users/benwibking/parthenon/build/Kokkos/core/src -I/Users/benwibking/parthenon/external/Kokkos/core/src -I/Users/benwibking/parthenon/external/Kokkos/core/src/../../tpls/desul/include -I/Users/benwibking/parthenon/build/Kokkos/containers/src -I/Users/benwibking/parthenon/external/Kokkos/containers/src -I/Users/benwibking/parthenon/build/Kokkos/algorithms/src -I/Users/benwibking/parthenon/external/Kokkos/algorithms/src -I/Users/benwibking/parthenon/build/Kokkos/simd/src -I/Users/benwibking/parthenon/external/Kokkos/simd/src -isystem /opt/homebrew/Cellar/open-mpi/5.0.3_1/include -isystem /Users/benwibking/ascent_build/install/ascent-develop/include -isystem /Users/benwibking/ascent_build/install/ascent-develop/include/ascent -isystem /Users/benwibking/ascent_build/install/conduit-v0.8.8/include -isystem /Users/benwibking/ascent_build/install/conduit-v0.8.8/include/conduit -isystem /Users/benwibking/ascent_build/occa/include -isystem /Users/benwibking/ascent_build/install/mfem-4.5.2/include -isystem /Users/benwibking/ascent_build/install/hdf5-1.14.1-2/include -isystem /Users/benwibking/ascent_build/install/raja-v2022.10.4/include -isystem /Users/benwibking/ascent_build/install/camp-2022.10.1/include -isystem /Users/benwibking/ascent_build/install/umpire-2022.10.0/include -isystem /Users/benwibking/ascent_build/install/vtk-m-v2.1.0/include/vtkm-2.1/vtkm/thirdparty/optionparser -isystem /Users/benwibking/ascent_build/install/vtk-m-v2.1.0/include/vtkm-2.1/vtkm/thirdparty/diy -isystem /Users/benwibking/ascent_build/install/vtk-m-v2.1.0/include/vtkm-2.1/vtkm/thirdparty/diy/vtkmdiy/include -isystem /Users/benwibking/ascent_build/install/vtk-m-v2.1.0/include/vtkm-2.1/vtkm/thirdparty/diy/vtkmdiy/include/vtkmdiy/mpi -isystem /Users/benwibking/ascent_build/install/vtk-m-v2.1.0/include/vtkm-2.1/vtkm/thirdparty/lcl/vtkmlcl -isystem /Users/benwibking/ascent_build/install/vtk-m-v2.1.0/include/vtkm-2.1 -isystem /Users/benwibking/ascent_build/install/ascent-develop/include/ascent/vtkh -isystem /Users/benwibking/ascent_build/install/ascent-develop/include/ascent/dray -isystem /Users/benwibking/ascent_build/install/ascent-develop/include/ascent/apcomp -Wno-deprecated-declarations  -O2 -g -DNDEBUG -std=c++17 -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -I/Users/benwibking/ascent_build/install/conduit-v0.8.8/include/conduit -DCONDUIT_RELAY_IO_MPI_ENABLED -DCONDUIT_BLUEPRINT_MPI_ENABLED -ffunction-sections -DROVER_PARALLEL -MD -MT src/CMakeFiles/parthenon.dir/bvals/comms/build_boundary_buffers.cpp.o -MF src/CMakeFiles/parthenon.dir/bvals/comms/build_boundary_buffers.cpp.o.d -o src/CMakeFiles/parthenon.dir/bvals/comms/build_boundary_buffers.cpp.o -c /Users/benwibking/parthenon/src/bvals/comms/build_boundary_buffers.cpp
In file included from /Users/benwibking/parthenon/src/bvals/comms/build_boundary_buffers.cpp:25:
In file included from /Users/benwibking/parthenon/src/bvals/comms/bvals_in_one.hpp:29:
In file included from /Users/benwibking/parthenon/src/tasks/tasks.hpp:37:
In file included from /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:19:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/future:1694:12: error: no viable conversion from returned value of type 'decltype(std::declval<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9) &>()())' (aka 'parthenon::TaskStatus') to function return type '(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24)'
    return _VSTD::__invoke(__f_.first(), _VSTD::forward<_ArgTypes>(__arg)...);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__config:715:17: note: expanded from macro '_VSTD'
#  define _VSTD std
                ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/future:1650:14: note: in instantiation of member function 'std::__packaged_task_func<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), std::allocator<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9)>, (lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) ()>::operator()' requested here
    explicit __packaged_task_func(_Fp&& __f) : __f_(_VSTD::move(__f), __default_init_tag()) {}
             ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/future:1763:32: note: in instantiation of member function 'std::__packaged_task_func<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), std::allocator<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9)>, (lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) ()>::__packaged_task_func' requested here
        ::new ((void*)&__buf_) _FF(_VSTD::forward<_Fp>(__f));
                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/future:1900:45: note: in instantiation of function template specialization 'std::__packaged_task_function<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) ()>::__packaged_task_function<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9)>' requested here
        explicit packaged_task(_Fp&& __f) : __f_(_VSTD::forward<_Fp>(__f)) {}
                                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__memory/shared_ptr.h:287:37: note: in instantiation of function template specialization 'std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) ()>::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), void>' requested here
        ::new ((void*)__get_elem()) _Tp(_VSTD::forward<_Args>(__args)...);
                                    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__memory/shared_ptr.h:995:55: note: in instantiation of function template specialization 'std::__shared_ptr_emplace<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) ()>, std::allocator<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) ()>>>::__shared_ptr_emplace<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9)>' requested here
    ::new ((void*)_VSTD::addressof(*__guard.__get())) _ControlBlock(__a, _VSTD::forward<_Args>(__args)...);
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__memory/shared_ptr.h:1004:19: note: in instantiation of function template specialization 'std::allocate_shared<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) ()>, std::allocator<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) ()>>, (lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), void>' requested here
    return _VSTD::allocate_shared<_Tp>(allocator<_Tp>(), _VSTD::forward<_Args>(__args)...);
                  ^
/Users/benwibking/parthenon/src/tasks/thread_pool.hpp:155:22: note: in instantiation of function template specialization 'std::make_shared<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) ()>, (lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), void>' requested here
    auto task = std::make_shared<std::packaged_task<return_t()>>(
                     ^
/Users/benwibking/parthenon/src/tasks/tasks.hpp:529:16: note: in instantiation of function template specialization 'parthenon::ThreadPool::enqueue<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24)>' requested here
          pool.enqueue([t, &ProcessTask]() { return ProcessTask(t); });
               ^
/Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'decltype(std::declval<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9) &>()())' (aka 'parthenon::TaskStatus') to 'const (lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) &' for 1st argument
          pool.enqueue([t, &ProcessTask]() { return ProcessTask(t); });
                       ^
/Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'decltype(std::declval<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9) &>()())' (aka 'parthenon::TaskStatus') to '(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:529:24) &&' for 1st argument
In file included from /Users/benwibking/parthenon/src/bvals/comms/build_boundary_buffers.cpp:25:
In file included from /Users/benwibking/parthenon/src/bvals/comms/bvals_in_one.hpp:29:
In file included from /Users/benwibking/parthenon/src/tasks/tasks.hpp:37:
In file included from /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:19:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/future:1694:12: error: no viable conversion from returned value of type 'decltype(std::declval<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9) &>()())' (aka 'parthenon::TaskStatus') to function return type '(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20)'
    return _VSTD::__invoke(__f_.first(), _VSTD::forward<_ArgTypes>(__arg)...);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__config:715:17: note: expanded from macro '_VSTD'
#  define _VSTD std
                ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/future:1650:14: note: in instantiation of member function 'std::__packaged_task_func<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), std::allocator<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9)>, (lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) ()>::operator()' requested here
    explicit __packaged_task_func(_Fp&& __f) : __f_(_VSTD::move(__f), __default_init_tag()) {}
             ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/future:1763:32: note: in instantiation of member function 'std::__packaged_task_func<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), std::allocator<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9)>, (lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) ()>::__packaged_task_func' requested here
        ::new ((void*)&__buf_) _FF(_VSTD::forward<_Fp>(__f));
                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/future:1900:45: note: in instantiation of function template specialization 'std::__packaged_task_function<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) ()>::__packaged_task_function<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9)>' requested here
        explicit packaged_task(_Fp&& __f) : __f_(_VSTD::forward<_Fp>(__f)) {}
                                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__memory/shared_ptr.h:287:37: note: in instantiation of function template specialization 'std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) ()>::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), void>' requested here
        ::new ((void*)__get_elem()) _Tp(_VSTD::forward<_Args>(__args)...);
                                    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__memory/shared_ptr.h:995:55: note: in instantiation of function template specialization 'std::__shared_ptr_emplace<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) ()>, std::allocator<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) ()>>>::__shared_ptr_emplace<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9)>' requested here
    ::new ((void*)_VSTD::addressof(*__guard.__get())) _ControlBlock(__a, _VSTD::forward<_Args>(__args)...);
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__memory/shared_ptr.h:1004:19: note: in instantiation of function template specialization 'std::allocate_shared<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) ()>, std::allocator<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) ()>>, (lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), void>' requested here
    return _VSTD::allocate_shared<_Tp>(allocator<_Tp>(), _VSTD::forward<_Args>(__args)...);
                  ^
/Users/benwibking/parthenon/src/tasks/thread_pool.hpp:155:22: note: in instantiation of function template specialization 'std::make_shared<std::packaged_task<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) ()>, (lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9), void>' requested here
    auto task = std::make_shared<std::packaged_task<return_t()>>(
                     ^
/Users/benwibking/parthenon/src/tasks/tasks.hpp:538:12: note: in instantiation of function template specialization 'parthenon::ThreadPool::enqueue<(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20)>' requested here
      pool.enqueue([t, &ProcessTask]() { return ProcessTask(t); });
           ^
/Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'decltype(std::declval<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9) &>()())' (aka 'parthenon::TaskStatus') to 'const (lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) &' for 1st argument
      pool.enqueue([t, &ProcessTask]() { return ProcessTask(t); });
                   ^
/Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'decltype(std::declval<(lambda at /Users/benwibking/parthenon/src/tasks/thread_pool.hpp:156:9) &>()())' (aka 'parthenon::TaskStatus') to '(lambda at /Users/benwibking/parthenon/src/tasks/tasks.hpp:538:20) &&' for 1st argument
2 errors generated.
```

**Comment by pbrady** [comment]
I guess it's not a complete drop-in replacement.  I've got a fix in #1124.
